GRW-58 / Feature / Offer Page Debugger for new Onboarding API

robinandeer opened the pull request:
## What?

Co-authored with @dendoesit.

Implement a new Offer Page Debugger that uses the new QuoteCart API: `/se-en/new-member/offer-debugger`.

The goal is _not_ to improve the debugger but the setup is simplified quite a bit:
- Create quote bundles rather than multiple quotes per request
- Skip session handling (we worry about that only in the real app, if we need to)
- Implement `ErrorBoundary` to deal with errors in a React-y way (I can revert this if no one likes it 😜)
- Skip UUID generation of quote IDs since API can return them for us 🎉 

> Currently, the debugger passes the quote cart ID to the offer page as a query parameter. That will work okay for now but we might need to go back and revise this. Let's see.

## Why?

The new debugger is implemented as a copy of the old one. This will make it easy to remove the old one when we are ready to do so. It also makes it easy to skip complicated stuff like dealing with auth tokens and sessions that we don't need to do with the new API.

A new dependency is added `react-error-boundary` to handle errors in a more React-y way. There seems to be some conflict with Sentry that catches errors without letting them bubble up normally. Does anyone know how that works? @styrelius @gustaveen?

**Referenced ticket(s): [GRW-583]**

## Screenshots / recordings

![Screenshot 2021-11-15 at 10 18 35](https://user-images.githubusercontent.com/1220232/141755175-5bf7df03-5fe4-48fe-91fd-56e325e8470a.png)

## Review app

Test it live: https://web-onboardi-grw-583-fe-hbvsqd.herokuapp.com/se-en/new-member/offer-debugger

[GRW-583]: https://hedvig.atlassian.net/browse/GRW-583?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Commit:
Add debugger page for new API

Copy old debugger setup.
Add new graphql mutation for create session.
Integrate new session create mutation in debugger.

Commit:
Remove quote ID related things from debugger

Change debugger prop session token => session ID.
Use new useCurrentLocale hook.

Commit:
Refactor QuoteData to create quotes

Add generic create quote mutation.
Update forms to support new API.

Commit:
Remove custom hook

Commit:
Move error handling to debugger

Add error boundary package.
Refactor if statement to switch statement.

Commit:
Update quote types

Update error handling.

Commit:
Fix type import

robinandeer commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+
+const Wrapper = styled.div`
+  max-width: 900px;
+  padding: 1rem;
```
I've been wanting to fix this forever 😅

robinandeer commented on src/client/pages/OfferDebugger/components/QuoteData.tsx:
```diff
@@ -0,0 +1,358 @@
+    }
+  }
+
+  if (submitError) {
+    throw submitError
+  }
```
It seems like Sentry catches all errors and just logs them to the console which isn't super helpful when you want to use error boundaries. That's why I had to come up with this workaround.

robinandeer commented on src/client/pages/OfferDebugger/components/QuoteFormSweden.tsx:
```diff
@@ -0,0 +1,65 @@
+  ssn: '199509291234',
+  startDate: '',
+  email: 'sven.svensson@hedvig.com',
+  swedishApartmentData: {
```
Here's one of the few changes in the new API: `swedishApartment` => `swedishApartmentData`

robinandeer replied:
This will use the generic api in the future but it's a little unclear how this will work at the moment so let's go with this for now.

styrelius replied:
OK 🙃

robinandeer commented on src/client/pages/OfferDebugger/components/QuoteFormSweden.tsx:
```diff
@@ -0,0 +1,65 @@
+    zipCode: '12345',
+    livingSpace: 23,
+    householdSize: 1,
+    subType: ApartmentType.Rent,
```
...and the other change in the new API: `type` => `subType`

Commit:
Update field isStudent => student

guilhermespopolin reviewed (commented):
LGTM. Great approach used to insert such change. One thing tho is that I was trying to test it locally but it keeps returning the following error on `QueryBundleVariants` query:

```
{
  "errors": [
    {
      "message": "invalid json response body at https://gateway.test.hedvig.com/underwriter/public/quotes? reason: Unexpected end of JSON input",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "quoteBundle"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "errors": [
            {
              "message": "invalid json response body at https://gateway.test.hedvig.com/underwriter/public/quotes? reason: Unexpected end of JSON input",
              "locations": [],
              "path": [
                "quoteBundle"
              ]
            }
          ]
        }
      }
    }
  ],
  "data": null
}
``` 
Maybe something change on the API 🤷🏻‍♂️

robinandeer commented:
@guilhermespopolin hmmm which quote are you trying to create? I would expect the API to be a little shaky but SE quotes have worked for me at least.

guilhermespopolin commented:
> @guilhermespopolin hmmm which quote are you trying to create? I would expect the API to be a little shaky but SE quotes have worked for me at least.

All right, jokes on me I guess. It happen when I try to go to offer page but that's because this part is not in the scope of this PR, right?! 🤦🏻
Tested again creating quotes of all types (Swedish Apartment, Swedish House and Swedish Apartment + Accident) and it works :)

Commit:
Update to use new quote cart API

Commit:
Update danish quote values

Commit:
Update norway quote values

guilhermespopolin commented on src/client/graphql/CreateQuoteBundle.graphql:
```diff
@@ -0,0 +1,17 @@
+mutation CreateQuoteBundle($id: ID!, $quotes: [QuoteData!]!) {
```
Do you think would be worth it this renaming?

```suggestion
mutation CreateQuoteBundle($quoteCardId: ID!, $quotes: [QuoteData!]!) {
```

robinandeer replied:
Right, yeah I can see that this is clearer.

guilhermespopolin commented on src/client/graphql/QuoteCart.graphql:
```diff
@@ -0,0 +1,96 @@
+    bundle {
+      possibleVariations {
+        id
+        tag(locale: sv_SE)
```
Wouldn't be the case of getting `locale` as an argument of `QuoteCart` query?

robinandeer replied:
haha yeah sorry, this is confusing -- I actually update this query in another PR! But maybe I should fix it here already

guilhermespopolin commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+export const Debugger: React.FC = () => {
+  const locale = useCurrentLocale()
+  const [
+    createOnboardingSession,
```
Just a question: Do we wanna stick with this name or what we understood as a _onboarding session_ is now a _quote card_?

robinandeer replied:
Oh what -- good catch! I tried to update it everywhere but must've missed this one

guilhermespopolin reviewed (dismissed):
LGTM :)
Just left a couple of comments

Commit:
Small tweaks

Use more explicit parameter to create quote bundle.
Pass locale to quote cart query.
Rename function to create quote cart.

Co-Authored-By: Guilherme S. Popolin <19200662+guilhermespopolin@users.noreply.github.com>

styrelius commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+)
+
+export const Debugger: React.FC = () => {
+  const locale = useCurrentLocale()
```
Maybe some destructuring? 😀 
```suggestion
  const locale = useCurrentLocale()
  const { apiMarket, isoLocale } = locale
```

styrelius commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+    createOnboardingQuoteCart,
+    { data },
+  ] = useCreateOnboardingQuoteCartMutation()
+  const quoteCartId = data?.onboardingQuoteCart_create
```
I know this is not anything that can be done in this PR of course but it feels kind of weird that the `onboardingQuoteCart_create` just returns an ID as a string and not a data structure like {id: [the quote cart ID]}. Because when using this mutation it's not clear what the data you get back is, unless you check the schema 🤔 But it's nice to not always have to do that, and just get help by VS Code instead.

robinandeer replied:
Sure, please comment that to Tobias or directly in https://github.com/HedvigInsurance/rfc/pull/7!

styrelius commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+      <h1>Offer Page Debugger</h1>
+
+      <Row>
+        <h3>Session: {quoteCartId}</h3>
```
Maybe be clearer on what the "Session" is to someone only using the debugger and not looking at the code? 🤔 
```suggestion
        <h3>Quote cart session ID: {quoteCartId}</h3>
```

robinandeer replied:
Yes indeed, let's stick to one concept as much as possible.

styrelius replied:
So should we change the `<h3>`?

gustaveen reviewed (dismissed):
Nicely done! Worked great for me, let's get this out the door 🙌

styrelius commented on src/client/pages/OfferDebugger/components/QuoteData.tsx:
```diff
@@ -0,0 +1,393 @@
+
+export const QuoteData: React.FC<OfferProps> = ({ quoteCartId }) => {
+  const currentLocale = useCurrentLocale()
+  const currentMarket = useMarket()
```
Can we instead use `useCurrentLocale` from `/l10n/useCurrentLocale.ts` and then pick the `market` property from  the returned `locale` object instead? 🙏 I'm trying to get rid of all these old locale/market utils functions (and yes, I should have marked them as deprecated but I didn't think about that 🙈)

guilhermespopolin reviewed (dismissed):
Forgot to approve hahah

styrelius commented on src/client/pages/OfferDebugger/index.tsx:
```diff
@@ -0,0 +1,21 @@
+  return { default: module.Debugger }
+})
+
+export const OfferDebugger: React.FC = () => {
+  if (window.hedvigClientConfig.appEnvironment === 'production') {
+    return null
+  }
```
I know this is what we do currently with the debugger but it's kind of weird that if you go to this route in prod you just get a black page. Could we redirect to the landing page instead? As we do with all other non-existent routes?

robinandeer replied:
Sure, I don't see why not 👍

styrelius reviewed (changes requested):
Great job 💐 ✨  

Maybe this is a golden opportunity to rename this whole feature something like **Quote Generator** or **Quote Creator**  which we've talked about previously ("debugger" is kind of a confusing name IMO)?

styrelius reviewed (commented):
OK I know we want to merge this so I'm changing this to an *Approve*. That said, I still want my comments answered and I'd still love for the `useMarket` hook to be replaced 🙏

styrelius commented:
Aha LOL you can't apparently just change your review from "Requesting changes" to "Approving". I'm so sorry 😔

robinandeer commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+      foreground={colorsV3.gray100}
+      onClick={resetErrorBoundary}
+    >
+      👉 Try starting over by creating a new session!
```
We are not storing any state outside React in this version of the debugger. Do you still think something is missing that needs to be reset?

styrelius replied:
No it was a joke, sorry 🙃

robinandeer commented:
> Aha LOL you can't apparently just change your review from "Requesting changes" to "Approving". I'm so sorry 😔

No worries, I'll review the changes and re-request review.

> Maybe this is a golden opportunity to rename this whole feature something like Quote Generator or * Quote Creator* which we've talked about previously ("debugger" is kind of a confusing name IMO)?

Sure, probably a good idea. However, like I wrote in the "What?" section I did not set out to make general improvements in this PR -- just make sure the debugger works with the new API!

robinandeer commented on src/client/graphql/CreateQuoteBundle.graphql:
```diff
@@ -0,0 +1,17 @@
+mutation CreateQuoteBundle($quoteCartId: ID!, $quotes: [QuoteData!]!) {
+  quoteCart_createQuoteBundle(id: $quoteCartId, input: { quoteData: $quotes }) {
+    ... on QuoteCart {
```
Yep, we can make that optimisation but don't think it will give us much in this debugger tool, do you agree?

I think we can assume everything is fine unless we get back `CreateQuoteBundleError`.

styrelius replied:
Yeah I agree maybe the "debugger" is not the place where those kinds of optimizations but it just feels strange and it's something I've been thinking about every time I've seen the debugger code 🤷‍♀️ But let's skip doing anything about that then.

robinandeer commented on src/client/pages/OfferDebugger/Debugger.tsx:
```diff
@@ -0,0 +1,88 @@
+    createOnboardingQuoteCart,
+    { data },
+  ] = useCreateOnboardingQuoteCartMutation()
+  const quoteCartId = data?.onboardingQuoteCart_create
```
You are referring to the API, right? I think it would be great to add that to https://github.com/HedvigInsurance/rfc/pull/7!

Commit:
Update locale utils

Add redirect from offer debugger in production.

Co-Authored-By: Ellen Styrélius <42962286+styrelius@users.noreply.github.com>

guilhermespopolin reviewed (approved):
![](https://media4.giphy.com/media/xHMIDAy1qkzNS/giphy.gif)

robinandeer committed:
Merge branch 'master' into GRW-583/feature/new-api-debugger

styrelius reviewed (approved):
👏
